src/generators.ts

```diff
@@ -43,3 +32,54 @@ export function last<T>(iter: Iterable<T>): T {
     }
     return val;
 }
+
+export class MemoizedIterable<T> {
+    private iterator: Iterator<T>;
+    private nextIndex: number;
+    private complete: boolean;
+
+    public constructor(iter: Iterator<T> | Iterable<T>) {
+        // case when iter is Iterable
+        if (iter[Symbol.iterator]) {
+            this.iterator = iter[Symbol.iterator]();
+        } else {
+            this.iterator = iter as Iterator<T>;
+        }
+        this.nextIndex = 0;
+        this.complete = false;
+    }
+
+    public get length(): number {
+        while (!this.complete) {
+            this.nextValue();
+        }
+        return this.nextIndex;
+    }
+
+    private nextValue(): T {
```

Comment:
This could be written as an generator:

```js
    private *consumeIterator(): Iterator<null> {
        for (const result of {[Symbol.iterator]: this.iterator}) {
            this[this.nextIndex++] = result.value;
            yield null;
        }
        this.complete = true;
    }
```

And store an instance in the constructor:

```js
        this.nextValue = this.consumeIterator().next;
```

Reply:
It could alternatively `yield;`, wasn't sure if it'd be ok to yield undefined. It could also echo the inner iterators value.